### DISCUSSION TITLE
[mini.completion] lag on cursor movement?

### DISCUSSION THREAD

**Discussion by pkazmier** [nvim-mini/mini.nvim#814] [Q&A] [answered]
Has anyone else noticed lag with the cursor position following the letters you quickly type in a buffer when there are possible completion candidates? It's a very odd effect. I see the characters displayed as I type them, but the cursor remains / lags behind and then will eventually (after 3-4  characters have been typed quickly) jump to the correct position as I'm typing. Very disconcerting effect.

I notice this only when typing a headline in a markdown file with the marksman LSP server—especially if the headline contains text already in the buffer.

I wonder if this could be due to slow LSP responses. I was going to try to write my own LSP server with explicit delays on responses to prove my theory, but thought I'd ask here beforehand.

It's easily repeatable on my work computer (loaded with all sorts of typical antivirus/DLP agents that make it crawl), but I haven't been able to recreate on my home Mac (way faster).

If I disable the LSP, the problem goes away. If I switch back to nvim-cmp, the problem does not happen. If I change the delay in mini completion to a higher value (5 seconds), the problem goes away.

Thanks.

**Comment by echasnovski** [comment]
This looks like a result of e3d13e2c89a9dfda51c69c825360d1fe5be0e322 which (accidentally) started forcing full two-step completion on LSP triggers. This is not correct and it should force only LSP completion without fallback (which is probably the culprit here as default `<C-n>` fallback is synchronous).

This particular issue should be fixed on latest `main`. Would you mind checking with your use case?

**Comment by pkazmier** [comment]
It did not. Here is a quick recording to show what I'm referring to if it helps.

https://github.com/echasnovski/mini.nvim/assets/747855/8a6c3a32-6a84-4978-9c08-a338f1de5c4a

Let me know what else I can do to help narrow this down for you.

**Reply by echasnovski** [reply, marked as answer]
This looks visually similar #678. If[ disabling fallback removes this issue](https://github.com/echasnovski/mini.nvim/issues/678#issuecomment-1925430272) (and judging by the video it would), then I don't think anything can be done here.

**Reply by pkazmier** [reply]
Yes, you are correct. Disabling fallback eliminates the issue.

Sorry, I didn't find that issue beforehand. I searched the discussions for lag/slow, but forgot to check issues.
